content/blog/2023-12-11-circ-bitwidth-adjustment

```diff
@@ -0,0 +1,115 @@
+Without our optimization, the cost spent on adjusting the bitwidth will be: 33 constraint to adjust `a + b` and `a + c` respectively, and 64 to adjust `c * d`, which result in 130 constraints in total.
+In contrast, after our optimization, the cost spent on adjusting `a + b` and `a + c` would be 0, and 66 to adjust `c * d`, which result in 66 constraints in total. 
+
+## Implementation
+1. We create a new type of operations: `add_unadjust` and `mul_unadjust` for addition and multiplication to notify the constraint generation system not to adjust the bitwidth.
+2. We implemented two passes to rewrite the operations. First pass is to classify elements in the system. We implemented an initial pass to scan the entire program and store elements requiring adjustments in a set S. Whenever we encounter an operation that definitely requires adjustment, we add the elements used in the operation to S. Then we implemented a rewrite pass to rewrite operations done on bitvectors not included in D to our newly introduced unadjust operation. 
+3. We have also made changes to the constraint generation system to take advantage of the two new operations. In the original implementation, the constraint generation system will always adjust the bitwidth of the result of an operation. We have made changes to the system to not adjust `add_unadjust` and `mul_unadjust` operations. We also proved that such change will not impact the correctness of the system.
+
+## What were the hardest parts to get right
+The hardest part is to make sure the compiler works properly using our newly introduced operation. We encounter many bugs while implementing the optimization. For instance, before since the author always adjust the bitwidth after all types of operations, the bitwidth is fixed. However, now since an operation might be using unadjusted bitwidth the logic of calculating the bitwidth of the current element is more complicated and requires more steps to ensure an operation does not overflow.
+
+## Empirical evaluation
+We uses two types of tests. 
+1. We use the existing benchmarks in the circ repo to test whether the output of the optimized compiler is correct. Our implementation passed all the test cases in CirC.
+2. To measure how well the proposed design improves the system, we use the number of r1cs constraints generated as the metrics. We runs our compiler and the original one over a set of zokrates programs in [CirC](https://github.com/circify/circ) and [Zombie](https://github.com/PepperSieve/Zombie/tree/master) and compares the number of constraints generated. Here is a shorted list of tests we run. The first line are the numbers of constraints with our flexible adjustment implementation and the second line is the result using original compiler.
+```
```

Comment:
Could you format this as a table or as a bar chart? Simply listing the text output from a tool is pretty hard to digest.